Express in vector module

szymag opened the pull request:
This PR adapt `express` for transformation equations.

szymag committed:
Apply changes

szymag committed:
Add compose function.

szymag committed:
Add compose function.

szymag commented:
ping @Upabjojr

Upabjojr commented on sympy/vector/functions.py:
```diff
@@ -5,6 +5,24 @@
+def compose(expr, system):
+    from sympy.vector.operators import _get_coord_sys_from_expr
+
+    x, y, z = symbols('x y z')
```
Use `lambda`, not symbols.

szymag committed:
Merge branch 'express' of https://github.com/szymag/sympy into express

szymag committed:
Merge branch 'constructor_fix' into express

szymag committed:
Revert direction of composition.

Update tests.
Transformation equations are are supported in express by scalars.

Upabjojr commented:
We need tests of `express` with the new coordinate transformations.

szymag committed:
Update direction of transformation in CoordSys3D and _compose

szymag committed:
Support transformation equations for vector expression in `express`

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -446,15 +446,18 @@ def lame_coefficients(self):
-        return self._transformation_lambda(*self.base_scalars())
+        if self._parent is None:
+            raise ValueError("No parent coordinate system, use `transformation_from_parent_function()`")
+        return self._transformation_to_parent_lambda(*self._parent.base_scalars())
```
why `self._parent.base_scalars()`? You are transforming the current base scalars to the parent's one.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -446,15 +446,18 @@ def lame_coefficients(self):
     def transformation_to_parent(self):
-        return self._transformation_lambda(*self.base_scalars())
+        if self._parent is None:
+            raise ValueError("No parent coordinate system, use `transformation_from_parent_function()`")
```
why? the parent technically exists even if it is `None`.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -446,15 +446,18 @@ def lame_coefficients(self):
+        return self._transformation_to_parent_lambda
 
     def transformation_from_parent(self):
-        if self._parent is None:
```
if the parent does not exist, we cannot derive an expression from it.

Upabjojr commented:
At the first glance, it looks like you have inverted the transformation functions twice, so that the effect has cancelled.

szymag commented:
> At the first glance, it looks like you have inverted the transformation functions twice, so that the effect has cancelled.

Can you show me the place which suggest that?

Upabjojr commented:
Can you undo the edits to the transformation functions?

szymag commented:
Yes

szymag committed:
Revert edits to the transformation equations.

Upabjojr commented:
I think it's simple, we are using the transformation input as transformation from the parent, while the code of vector used to assume that it is the opposite.

@moorepants hasn't yet answered to my email.

Upabjojr commented:
One possible interpretation is that the vector module is tought to manage transformations as active transformations: instead of rotating the coordinate system, you are rotating all objects in the coordinate system. Therefore the result is like having the coordinate system rotated by the opposite transformations.

szymag commented:
http://docs.sympy.org/latest/modules/vector/coordsys.html#orienting-new-systems
There is a picture which suggest that we have here passive rotation. Also for active we have separate method `orient_new_body`

szymag committed:
Correct sign for location, transpose jacobian

szymag committed:
Merge branch 'master' into express

szymag committed:
Finsh express reconstruction.

Fix _projections method in Vector class.
Coord_sys were sometimes taken not for vector but for scalars.

szymag committed:
Add test_transformation_functions

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -482,3 +482,23 @@ def test_translation_trans_equations():
+    assert express(E.x, C, variables=True) == C.x*cos(q1) + C.y*sin(q1)
+    F = C.create_new("D", lambda x, y, z: (x**3, y, z))
+    F._calculate_inv_trans_equations()
+    assert F.transformation_from_parent() == (C.x**(1/3), C.y, C.z)
```
1/3 needs to be S(1)/3, otherwise it won't be a rational.

Upabjojr replied:
By the way, shouldn't it be the opposite direction?

Upabjojr replied:
Try to also add one test with `lambda x, y, z: x+3, y, z`. This one should behave exactly like the locate_new one.

szymag replied:
> By the way, shouldn't it be the opposite direction?
 
I think that direction is OK. Transformation equations defines transformation *from* parent

szymag replied:
*to* parent should be inverse transformatin. I think that we need to exchange names in `transformation_to_parent` `transformation_from_parent`

Upabjojr replied:
The transformation defines `F.x == C.x**3`, right? _C_ is the parent so `F.x = C.x**3` is from parent, while `C.x = F.x**(S(1)/3)` is from parent (by inferring the equation). Right?

szymag replied:
> The transformation defines F.x == C.x**3, right?

I'd say that it should be C.x == F.x**3.
https://en.wikipedia.org/wiki/Spherical_coordinate_system#Cartesian_coordinates
On the LHS of transformation equations we have variables from older coordinate system, in that case from parent.
On the LHS of inverse transformation equations we have variables from newer coordinate system.

Upabjojr replied:
Well... we agreed that `transformation` applies to the coordinate system, right?
I would expect for `spherical` to have:
```python
>>> G = C.create_new("G", "spherical", variable_names=["r", "theta", "phi"])
>>> G.transform_from_parent()
(sqrt(C.x**2 + C.y**2 + C.z**2), ... )
>>> G.transform_to_parent()
(G.r*sin(G.theta)*cos(G.phi), ... )
```
If we want to specify a passive transformation, we should use:
```python
>>> G = C.create_new("G", lambda x, y, z: (sqrt(x**2 + y**2 + z**2), ... , ... ))
```
I believe we have so far done the exact opposite.

That means, if we want to have a spherical transformation and keep a passive transformation convention (not active), we should give the inverse transformation to spherical to `transform` to get a spherical transformation.

`transformation` may have various forms:

- rotation matrix + tranlsation vector, 
- lambda,
- equation mapping,
- string.

They should all be either passive or active transformation definitions (let's avoid mixing the two).

Given that the previous code uses passive transformations, we should keep the passive convention, i.e. the transformation specifies how the coordinate system gets transformed, while the vectors remain at the same position.

szymag committed:
Apply passive transformation

szymag committed:
Update tests, transformation methods.

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -141,8 +141,8 @@ def __new__(cls, name, transformation=None, parent=None, location=None,
-            lambda_inverse = lambda x,y,z: r.inv()*Matrix(
-                [x-l[0], y-l[1], z-l[2]])
+            lambda_inverse = lambda x,y,z: tuple(CoordSys3D._inverse_rotation_matrix(r)*Matrix(
+                [x-l[0], y-l[1], z-l[2]]))
```
these should be `+` in the passive transformation, right?

Upabjojr commented on sympy/vector/coordsysrect.py:
```diff
@@ -445,15 +446,17 @@ def lame_coefficients(self):
 
     def transformation_to_parent(self):
-        return self._transformation_lambda(*self.base_scalars())
+        return self._transformation_from_parent_lambda(*self.base_scalars())
```
is this a name mismatch? When I suggested to invert the transformation directions, I meant to keep the name consistent here.

Upabjojr replied:
`self._transformation_from_parent_lambda` should be in `transformation_from_parent`.

szymag replied:
I also saw this problem. But only in that configuration test will pass

Upabjojr replied:
Name consistency please.

Upabjojr committed:
Update coordsysrect.py

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,29 @@ def test_translation_trans_equations():
+                    r * sin(theta) * sin(phi),
+                    r * cos(theta)))
+    assert raises(ValueError, lambda: F.transformation_from_parent())
+    assert F.transformation_to_parent() == (G.x*sin(G.y)*cos(G.z), G.x*sin(G.y)*sin(G.z), G.x*cos(G.y))
```
add some tests with the transformation to spherical using both 
-`transformation="spherical"`, 
- another test using `lambda x, y, z: (sqrt(x**2 + y**2 + z**2), ... , ... )`.
- `transformation=((x, y, z), (x + 3, y, z))` ==> this should behave like the similar case above.

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,29 @@ def test_translation_trans_equations():
+    assert F.transformation_to_parent() == (F.x**3, F.y, F.z)
+    assert express(C.x, F, variables=True) == F.x**(1/3)
+    assert express(F.x, C, variables=True) == C.x**3
+    G = C.create_new('G', lambda r, theta, phi: (
```
we concluded that this is the inverse transformation, right? That is, we should use this definition if the parent is spherical, because we are in a passive transformation definition. `transformation_from_parent` will be the inverse of `lambda` while `transformation_to_parent` will be `lambda`. Am I correct?

Upabjojr committed:
reversed transformation directions

szymag committed:
Update missing transformation

szymag committed:
Update constructor and tests

szymag commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,36 @@ def test_translation_trans_equations():
+    print(lambda: G.transformation_from_parent())
+    #assert raises(ValueError, lambda: G.transformation_from_parent())
+    assert G.transformation_to_parent() == (G.x*sin(G.y)*cos(G.z), G.x*sin(G.y)*sin(G.z), G.x*cos(G.y))
+    H = C.create_new("H", lambda x, y, z: (x-3, y, z))
```
I believe this represents the same transformation as `C.locate_new("D", 3 * C.i)`.
`to_parent` shows how `parent` is modified and it should be shifted to the left.

szymag commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,35 @@ def test_translation_trans_equations():
+                    r * sin(theta) * cos(phi),
+                    r * sin(theta) * sin(phi),
+                    r * cos(theta)))
+    # assert raises(ValueError, lambda: G.transformation_from_parent())
```
This test fails on my PC. I don't know why because `ValueError` is raised

Upabjojr replied:
Do not comment if it's the expected value. So we can see Travis.

Upabjojr replied:
You don't need `assert` if you use `raises`

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,35 @@ def test_translation_trans_equations():
+def test_transformation_functions():
+    C = CoordSys3D("C")
+    D = C.locate_new("D", 3 * C.i)
+    assert D.transformation_from_parent() == (C.x + 3, C.y, C.z)
```
didn't we conclude that `D.x == C.x - 3`? Express replaces `C.x` with `D.x + 3` in line 492...

szymag replied:
If we have passive transformation, the `parent` coordinate system is shifted. Here is shifted in left direction, I mean `-3`.
So if we have `from_parent` we're gong to the right direction, `+3`.
It's my explanation. I might be wrong

Upabjojr replied:
Well, my purpose is to make `transformation_from_parent` return the combination of the parent's base scalars such that will be equivalent to the child's base scalars. Maybe the name was a bit unfortunate.

The idea is that you can use the resulting tuple and replace the value for `D.x` in `express`.

Maybe we should rename it to something more intuitive, like `base_scalars_from_parent`.

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,35 @@ def test_translation_trans_equations():
+                    r * cos(theta)))
+    # assert raises(ValueError, lambda: G.transformation_from_parent())
+    assert G.transformation_to_parent() == (G.x*sin(G.y)*cos(G.z), G.x*sin(G.y)*sin(G.z), G.x*cos(G.y))
+    H = C.create_new("H", lambda x, y, z: (x-3, y, z))
```
this case should be equivalent to `C.locate_new("H", -3*C.i)` or to `C.locate_new("H", 3*C.i)`?

szymag replied:
I guess you missed my comment:
> I believe this represents the same transformation as `C.locate_new("D", 3 * C.i)`.
to_parent shows how parent is modified and it should be shifted to the left.

Upabjojr replied:
Well, so far I meant to use _to parent_ to represent how the variables of the child should be modified to be equivalent to the variables of the parent.

Upabjojr commented on sympy/vector/tests/test_coordsysrect.py:
```diff
@@ -479,3 +482,35 @@ def test_translation_trans_equations():
+    assert F.transformation_to_parent() == (F.x**3, F.y, F.z)
+    assert express(C.x, F, variables=True) == F.x**(1/3)
+    assert express(F.x, C, variables=True) == C.x**3
+    G = C.create_new('G', lambda r, theta, phi: (
```
If C is Cartesian, this is the wrong transformation direction to give a spherical child.

szymag committed:
Add tests

szymag committed:
Merge branch 'master' into HEAD

czgdp1807 commented:
@szymag Are you still working on it? Thanks for the contribution.

czgdp1807 commented:
@moorepants @Upabjojr  Do we need this PR? I was thinking to attach a `Please take over` label to it.

moorepants commented:
I don't know about this one. Looks like @Upabjojr was more involved with it.

czgdp1807 commented:
@Upabjojr Please let us know if this is needed.

Upabjojr commented:
This PR is probably not going to be merged by now.
